service/src/java/org/apache/hive/service/cli/session/HiveSessionImpl.java

```diff
@@ -432,6 +432,18 @@ private synchronized void releaseBeforeOpLock(boolean userAccess) {
       sessionState.resetThreadName();
     }
 
+    // We have already set the thread-local Hive belonging to the current session,
+    // if the thread-local Hive has been changed/updated after running the operation,
+    // the Hive after should belong to the same session, and we should update the sessionHive.
+    // The thread-local hive would be recreated only when the underlying
+    // HiveMetaStoreClient is incompatible with the newest session conf.
+    Hive localHive = Hive.getThreadLocal();
+    if (localHive != null && localHive != sessionHive) {
+      // The previous sessionHive would be GC'ed finally, or should we force close it?
+      sessionHive = localHive;
+      sessionHive.setAllowClose(false);
```

Comment:
isn't this flag already set during open() in SessionState.getHiveDb()?

Reply:
Not yet, when `localHive != sessionHive` comes true, means that the thread-local hive is a fresh new instance, 
https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java#L428-L466

TestJdbcDriver2#testResultNextAcidTable can repro the case.